Bdgtc

steveri opened the pull request:
This branch produces an LVS-clean chip with dragonphy2 instead of dragonphy1, see https://buildkite.com/tapeout-aha/fullchip/builds/199 and r7arm-aha:/build/gold.199
```
    % ssh r7arm-aha
    % cd  /build/gold.199/full_chip
    % /home/steveri/tmpdir/buildcheck.sh/build/gold.199/full_chip

    +++ SIZE CHECK
    17-tile_array/17-Tile_PE       SIZE   90 BY   88 ;
    17-tile_array/16-Tile_MemCore  SIZE  279 BY   88 ;
    17-tile_array                  SIZE 4451 BY 1632 ;
    14-glb_top                     SIZE 2490 BY 1974 ;

    +++ LVS CHECK
      CORRECT        glb_tile                      glb_tile
      CORRECT        global_buffer                 global_buffer
      CORRECT        global_controller             global_controller
      CORRECT        Tile_PE                       Tile_PE
      CORRECT        Tile_MemCore                  Tile_MemCore
      CORRECT        Interconnect                  Interconnect
      CORRECT        GarnetSOC_pad_frame           GarnetSOC_pad_frame

```

Among the changes you will see vs. master:
 - enabled a script to automatically remove INCLUDE paths from dragonphy spice deck
 - commented-out a bunch of commands in garnet synthesis and/or layout scripts
 - disabled several mflowgen error checks b/c of "read_sdc" problems
 - reduced clock freq back to 250MHz
 - added scripts to automatically remove shorts after glb and pe tile synthesis (!)
 - added multiple checks to ensure correct tclsh versioning
 - changed default adk view from "stdview" to "multivt"
 - plus of course all the necessary changes to support dragonphy2 vs. dragonphy

In particular, this branch uses an aham3soc branch that has been updated to support dragonphy2, see "recommended procedure" below.

Also note this branch uses Steve's version of the dragonphy2, in particular dragonphy_top.gds has been hacked to bring some lower level ports up to the top level for black boxing; at some point we may want to coordinate bringing these in line with the "official" versions in `/home/ajcars/dragonphy_top/`; again see "recommended procedure" below.

There are a lot of changes here and a lot of things to coordinate, so I propose doing the update in several discrete stages, the first of which results in a clean Friday Night Build (finally!), while each successive stage fixes the various hacks that got us there.

Recommended procedure:
- finalize this garnet pull so as to get Friday Night Builds working again ASAP;
- *finally* update aham3soc master to support new dragonphy2 (https://github.com/StanfordAHA/AhaM3SoC/pull/8);
- file a new pull that updates garnet master to use aham3soc master instead of dev branch;
- coordinate with PHY guys to produce a clean black-boxable dragonphy_top.gds and add that to /home/ajcars repository
- file a new pull that updates garnet master to use /home/ajcars dragonphy instead of /sim/steveri

steveri commented:
Also see conversation in previous pull attempt https://github.com/StanfordAHA/garnet/pull/716

alexcarsello commented:
Steve,
I'll go through these changes in detail, but at first glance, I'm pretty confident that a lot of them are not necessary. Last night I ran a tile array on the master branch and it was not too large to fit on the chip. (See results at /sim/ajcars/fc-12-16 on r7arm-aha2). I'm not sure what was different in your run that caused the tile array size to increase.

steveri commented:
@alexcarsello I got the tile array back down to its original size before filing this pull, it fits just fine now...thought I sent out mail about that, I will check and resend (also see "SIZE CHECK" info above).

2) You will notice that most of the commented-out code is commented out *only for buildkite,* and stays commented-in otherwise. I think it will take significant effort to get it all commented back in such that there are no buildkite errors, so I was happy to postpone that in favor of having a working branch with the new dragonphy...

3) multivt vs. stdview shouldn't really change anything, right? It checks std libs first and if it finds the lib it wants it uses it, yes?

4) 250MHz is a godsend for CI, whose main job is to quickly find out if things are badly broken. It might be better to leave it up to the individual developers to make sure the final build runs at speed and is DRC clean. If it makes you feel better, I can make this also be a buildkite-only mod.

5) You guys are supposed to be on break! :)
I'm more than happy to wait until January to finalize all of this...meanwhile I just wanted to get it in the pipeline before it got stale on my end...

steveri commented:
@alexcarsello This is a portion of the email I sent recently (yesterday maybe?), you probably haven't seen it yet...

> One thing that bothered me about build 198 was the size of the PE tile, which for some reason was wider than when it gets built in the master branch. So I tweaked a couple of things and tried again. The retry managed to get the PE tile size back down to its original 90u width, see build 199 details below. I plan to file a git pull on this branch when/if I have a moment, so then people will have something fun to do when they get back from winter break... :)

alexcarsello commented:
Coming back to this...

On 3., I agree that changing to multivt is unlikely to break anything, but what are you trying to solve by changing it?

On 4, perhaps that's true of the CI we perform on each PR, but there's nothing quick about the weekly full chip builds. We want those to be as close to shippable as possible. As we saw during the tapeout, running the flow at higher clock speeds frequently uncovers issues that aren't present at low speeds, so I don't think we should be running all of our CI at 250 MHz until the very end.

alexcarsello commented on mflowgen/common/scripts/fix-shorts.tcl:
```diff
@@ -0,0 +1,132 @@
+}
+
+########################################################################
+# Delete route blockages (will restore them again later)
+echo "@file_info: Deleting route blockages"
+deleteRouteBlk -all
+redraw; sleep 1
+
```
What's the reason for getting rid of the blockages temporarily? This could cause the router to violate the blockages while trying to fix the shorts... was it unable to fix shorts with the blockages in place?

steveri replied:
Correct. Blockages existed that prevented the shorts from being fixed. After the blockages are restored, drc is re-checked to verify that we didn't re-route through a forbidden area.

alexcarsello commented on mflowgen/common/scripts/fix-shorts.tcl:
```diff
@@ -0,0 +1,132 @@
+        -name $blockage_names($i)   \
+        ;
+}
+redraw; sleep 1
```
Just a general comment on this script... I really like the idea of having an automated way of fixing shorts in innovus! It seems like this could be fairly easily extended to fix all metal geometry DRCs, not just shorts. That can be left for a future PR though.

steveri replied:
Agreed. But I think it's crazy that we have to do this. Innovus should be smarter. And it definitely shouldn't leave us with shorted wires without throwing some kind of error or at least a warning!!

alexcarsello commented on mflowgen/common/soc-rtl-v2/run.sh:
```diff
@@ -6,6 +6,10 @@ CUR_DIR=$(pwd)
 
+# We'll need this until the new PHY RDL is incorporated and working (right?);
+# then we can delete it before the final merge(s)
+(cd aham3soc; git checkout phy_pads)
```
Since we have all the collateral from them now, we should put it in place and merge the phy_pads branch.

steveri replied:
I see that you approved my aham3soc merge finally :)
So I will undo this hack and start a run to verify that it works (see upcoming to-do list).

alexcarsello commented on mflowgen/full_chip/construct.py:
```diff
@@ -21,7 +21,12 @@ def construct():
-  adk_view = 'stdview'
+  adk_view = 'multivt'
+
+  # [steveri 11/2020] adk_view was: stdview. Now: multivt (above).
+  # Really only need 'multivt' for lvs (dragonphy uses lvt/ulvt)
+  # but I'm not sure the right way to do that
+  # so oh well everybody gets multivt...
```
Ah, I see... it's to make LVS pass without hacking. I think that's reason enough. We should check the full_chip reports to see how many LVT/ULVT cells are used. I suspect it's not too many, but if it's a lot, we should probably come back to this.

steveri replied:
It's using a couple dozen LVT cells IIRC. But I have a better fix, as you probably have already noticed by now...

alexcarsello commented on mflowgen/full_chip/dragonphy/outputs/dragonphy_RDL.gds:
```diff
@@ -1 +1 @@
-/home/ajcars/dragonphy_top/dragonphy_RDL.gds
\ No newline at end of file
+/sim/steveri/dragonphy/Dec_2020/dragonphy_RDL.gds
```
These should be in /home, not /sim so that the fullchip flow works on both arm and arm2 without maintaining separate versions.

steveri replied:
Yes, of course. Would you like to copy them to /home/ajcars so as to preserve the existing convention(s)? Then we can change the links

alexcarsello replied:
just copied

steveri replied:
Updated the dragonphy links, they once again point to /home/ajcars instead of /sim/steveri, so this task is done.

alexcarsello reviewed (changes requested):
See comments

steveri commented:
> 
> 
> Coming back to this...
> 
> On 3., I agree that changing to multivt is unlikely to break anything, but what are you trying to solve by changing it?
> 
> On 4, perhaps that's true of the CI we perform on each PR, but there's nothing quick about the weekly full chip builds. We want those to be as close to shippable as possible. As we saw during the tapeout, running the flow at higher clock speeds frequently uncovers issues that aren't present at low speeds, so I don't think we should be running all of our CI at 250 MHz until the very end.

@alex addressing our three main concerns
- commented-out code
- multivt vs. stdview
- slow clock

Taking them each in order...

* Commented-out code: Since filing the original pull I have made various tweaks to this branch as well as to the mflowgen master branch, and no longer need the commented out code; if you compare the latest version to the original you'll see that the original commands have been restored.

* multivt vs. stdview: This is being addressed by https://github.com/mflowgen/mflowgen/pull/94 . After the merge is finalized, then we can use the simple change in garnet branch "bdgtc_stdview" to fix this. Also see https://github.com/StanfordAHA/garnet/issues/752

* The slow clock can be a separate topic...for now I'm willing to put Tile_PE back to its original speed so that we can move things along...look for that to happen in the very near future (today or tomorrow hopefully).

steveri commented on mflowgen/common/scripts/fix-shorts.tcl:
```diff
@@ -0,0 +1,132 @@
+
+########################################################################
+# Fix shorts 1: Fetch a fresh set of violation markers
+# Note Tile_PE has VDD shorts on lower layers that we choose to ignore...
+# FIXME what's up with the VDD shorts!!??
+clearDrc
+verify_drc -layer_range { M4 M9 }
```
Correct. The short(s) that were causing problems were on metal level M4 and/or above. Setting M4-M9 solved the problem, and I didn't want to iterate to see if I could narrow it down to the exact necessary range.

steveri replied:
So are we good on this one?

steveri commented on mflowgen/common/scripts/fix-shorts.tcl:
```diff
@@ -0,0 +1,132 @@
+
+    # Fix shorts 5: Check your work
+    clearDrc
+    verify_drc -layer_range { M4 M9 }
```
Again, I found a range (M4-M9) that solved my immediate problems...given the long turnaround time to verify changes, I didn't want to refine it further...

steveri replied:
So...can we mark this "resolved?"

steveri commented:
My new to-do list:
1. restore Tile_PE to its original frequency and verify that it will work

2. remove hard-coded coords from three locations in mflowgen/common/custom-power-chip/outputs/power-strategy-dualmesh.tcl

3. undo aham3soc hack now that we have merged with master

4. wait for alex to copy /sim/steveri dragonphy collateral to /home/alex; then update symlinks to reflect the change...

Also I should probably figure out why we're failing the buildkite tests... :(

alexcarsello commented:
The latest dragonphy_top collateral is now in /home/ajcars/dragonphy_top

steveri commented:
> My new to-do list:
> 
>     1. restore Tile_PE to its original frequency and verify that it will work
> 
>     2. remove hard-coded coords from three locations in mflowgen/common/custom-power-chip/outputs/power-strategy-dualmesh.tcl
> 
>     3. undo aham3soc hack now that we have merged with master
> 
>     4. wait for alex to copy /sim/steveri dragonphy collateral to /home/alex; then update symlinks to reflect the change...
> 
> 
> Also I should probably figure out why we're failing the buildkite tests... :(

I fixed the PE frequency-setting code so that the default action is for buildkite to use the default clock period instead of the slower 250MHz hack. So at this point I think we've addressed items in the to-do list 1, 3, 4 above...next I'll work on the final item 2.

steveri commented on mflowgen/common/scripts/fix-shorts.tcl:
```diff
@@ -0,0 +1,132 @@
+set shorts [dbGet top.markers { .subType eq "Metal Short" }]
+if { $shorts == "0x0" } {
+    echo "@file_info: No shorts found"
+} {
```
if { cond } { then-clause } { else-clause }, yes? No "else" or "then" keywords, right?

alexcarsello replied:
I've never heard of this syntax. Using else definitely works and is preferred so it's clear what is going on.

steveri replied:
> I've never heard of this syntax. Using else definitely works and is preferred so it's clear what is going on.

OMG will this merge never end.

Requesting changes like this on working code is not very agile, especially considering that each set of changes is another 32 hours of testing before we get another green light.

The more agile approach is to approve the working code and then file an issue on whatever cleanup is needed. This comment would apply equally to most of your comments that I have addressed so far, e.g. working PHY stripes that should be parameterized instead of hardcoded.

As a result this merge is becoming more and more unwieldy; I vote that *as soon as* we have working code again (see below), we merge immediately and start fresh with a new pull to address any lingering concerns...

Meanwhile it looks like our window has closed again; the latest peak/lassen/magma changes have rendered the PE tile too large again, I will be sending out mail about that soon.

steveri replied:
> I've never heard of this syntax. Using else definitely works and is preferred so it's clear what is going on.

Now that my rant is out of the way I can address your actual concern. Originally I thought you were mistaken about needing an "else", since my code works, and then I did multiple tests in a tcl shell to confirm that the syntax I was using was doing what I thought it was doing.

Having read the tck/tk manual page https://www.tcl.tk/man/tcl/TclCmd/if.htm I now see that both forms are acceptable. I agree that having the "else" is more readable, and I will try to remember to include it here and elsewhere.

It seems odd to use the "else" without using the "then", but I scanned through all our tcl code so far and that does indeed seem to be the style.

steveri commented on mflowgen/common/scripts/fix-shorts.tcl:
```diff
@@ -0,0 +1,132 @@
+    set shorts [dbGet top.markers { .subType eq "Metal Short" }]
+    if { $shorts == "0x0" } {
+        echo "@file_info: All shorts fixed"
+    } {
```
if { cond } { then-clause } { else-clause }, yes? No "else" or "then" keywords, right?

steveri commented on mflowgen/common/scripts/fix-shorts.tcl:
```diff
@@ -0,0 +1,132 @@
+        echo "@file_info: All shorts fixed"
+    } {
+
+        # PE requires two tries maybe
+        echo "@file_info: Oops looks like I failed; lemme try again"
+
+        # Fix shorts 3a: Show how many shorts were found
+        set nshorts [llength $shorts]
+        echo "@file_info: (2)Found $nshorts short circuit(s):"
+        echo [ dbGet [dbGet top.markers { .subType eq "Metal Short" }].message ]
+
+        # Fix shorts 4a: See if globalDetailRoute can fix the shorts with eco
+        echo "@file_info: (2)Fixing short circuits"
+        setNanoRouteMode -routeWithEco true
+        setNanoRouteMode -drouteEndIteration 2
+        globalDetailRoute
+
+        # Fix shorts 5a: Check your work
+        clearDrc
```
Agile methodology: 1) do what works 2) fix what's broken; we can add this as FIXME/TODO/ISSUE if you like.

steveri commented:
@alexcarsello I think all your concerns have been addressed (finally!), please check and see if you like what we've got now. Note that if we pull this off it will be the first/only time that the master branch successfully runs our latest code base including dragonphy2.

The chip builds with error and final result is LVS correct, also see
https://buildkite.com/tapeout-aha/fullchip/builds/210
`r7arm-aha:/build/gold.210/full_chip/`

There appear to be problems with some of the other CI tests, but they don't seem to be related to this branch (see recent mail I sent to the aha group).

steveri commented on mflowgen/common/custom-power-chip/outputs/power-strategy-dualmesh.tcl:
```diff
@@ -199,6 +199,13 @@ addStripe -nets {VSS VDD} -layer $pmesh_bot -direction horizontal \
+createRouteBlk \
+    -layer M8 \
+    -box {1723.075 4098.0 2833.095 4798.416} \
+    -name dragonphy_M8   \
```
Done.

steveri commented on mflowgen/common/custom-power-chip/outputs/power-strategy-dualmesh.tcl:
```diff
@@ -237,10 +244,22 @@ foreach i { 0 1 2 3 4 5 6 7 8 } {
 
 # Delete the two stray wires where pins are missing at the top right.
-set a { 50 4762 1360 4768 }
+set a { 2830 4762 4900 4768 } ; # Top *right*
```
Done.

steveri commented on mflowgen/common/custom-power-chip/outputs/power-strategy-dualmesh.tcl:
```diff
@@ -237,10 +244,22 @@ foreach i { 0 1 2 3 4 5 6 7 8 } {
+# FIXME (along with much of the code above...)
+deselectAll
+set a {1703.075 4098.0 2853.095 4798.416}
+editSelectVia -net {VSS VDD} -cut_layer VIA8 -area $a
```
Done.

steveri commented:
> @alexcarsello I think all your concerns have been addressed (finally!), please check and see if you like what we've got now. Note that if we pull this off it will be the first/only time that the master branch successfully runs our latest code base including dragonphy2.
> 
> The chip builds with error and final result is LVS correct, also see
> https://buildkite.com/tapeout-aha/fullchip/builds/210
> `r7arm-aha:/build/gold.210/full_chip/`
> 
> There appear to be problems with some of the other CI tests, but they don't seem to be related to this branch (see recent mail I sent to the aha group).

@alexcarsello all tests are passing now, i think we have a window in which we can finally complete this pull...please let me know if you have any lingering concerns...
